### DISCUSSION TITLE
Allow prefix class types as arguments for transient-scope

### DISCUSSION THREAD

**Discussion by aurtzy** [magit/transient#334] [Ideas]
Would it be a good idea to allow prefix classes/types as arguments in addition to a specific command (or list of)?  My concern is that it is not straightforward to extend if I want other prefixes (not necessarily defined in the package) to be accepted as well.

For example, I have defined a prefix class with a custom `transient-init-scope` method.  The expectation is that any prefixes of this class/type will use the format from this `transient-init-scope`.  It thus seems desirable to allow `transient-scope` to return a non-nil value if a prefix matches that type.

Alternatives for this use case that don't require changing the current implementation:

-   Define a list variable for "compatible prefixes" that can be added to by the user or external packages.
-   It might also work to add some hook/method-with-qualifier to the prefix class that automatically does this.

**Comment by tarsius** [comment]
I considered that case but decided against implementing it because it would complicate the documentation and likely be rarely used. I've implemented (but not tested) it now, see the `334` branch.

It would help me to decide whether this should be merged if you could implement your prefix commands without using this feature and then add an additional commit that demonstrates how that could be simplified by using this feature.

**Reply by aurtzy** [reply]
Sure!

I had a go at the second alternative I mentioned ("add some hook/method-with-qualifier to the prefix class that automatically does this [manage a list variable]").  The implementations can be found in the `scope-object-with-transient-334` branch ([link here](https://github.com/aurtzy/disproject/tree/scope-object-with-transient-334)), with the last commit making use of the feature from `334`.

(`scope-object` was an existing refactor branch with changes to scope that also make this possible, but the last two commits in `scope-object-with-transient-334` are the only relevant ones in this context)

Barring any bugs or polish, I do consider the without-`334` implementation a pretty minimal and satisfactory solution, so I don't feel strongly about including this feature if it's deemed too niche :)


## Example of usage

The most useful effect of doing this is that I (or a user) can define prefixes like the following:

```emacs-lisp
(transient-define-prefix foo-project-dispatch ()
  :class disproject-prefix
  ["Commands"
   ("f" "Foo" foo)])

(transient-define-suffix foo ()
  (interactive)
  ;; This macro applies variables from the scope.
  (disproject-with-environment
    (message "Executed foo in the selected project!")))
```

&#x2026;which lets my `transient-init-scope` method automatically set up scope with a value that can also come from - or be passed to - other prefixes of the same type.

**Reply by tarsius** [reply]
I'll probably add this, but it won't make it into the next release, scheduled for tomorrow, because I am still unsure about some details.

**Reply by tarsius** [reply]
I've merged it. I went with the approach that uses two separate arguments.
